ipn/ipnlocal/serve.go

```diff
@@ -37,6 +42,166 @@ type serveHTTPContext struct {
 	DestPort uint16
 }
 
+type serveListener struct {
+	ap     netip.AddrPort
+	ctx    context.Context    // valid while listener is desired
+	cancel context.CancelFunc // for ctx, to close listener
+	logf   logger.Logf
+
+	ln net.Listener     // non-nil if listener is active
+	bo *backoff.Backoff // for retrying failed Listen calls
+}
+
+func newServeListener(ctx context.Context, ap netip.AddrPort, logf logger.Logf) *serveListener {
+	ctx, cancel := context.WithCancel(ctx)
+	return &serveListener{
+		ap:     ap,
+		ctx:    ctx,
+		cancel: cancel,
+		logf:   logf,
+
+		bo: backoff.NewBackoff("serve-listener", logf, 30*time.Second),
+	}
+}
+
+func (s *serveListener) Close() error {
+	s.cancel()
+	return s.ln.Close()
+}
+
+func (s *serveListener) Run(h func(net.Conn)) {
+	for {
+		if s.ctx.Err() != nil {
+			// context cancelled before we could listen
+			return
+		}
+		ln, err := net.Listen("tcp", s.ap.String())
+		if err != nil {
+			s.logf("serve failed to listen on %v, backing off: %v", s.ap, err)
+			s.bo.BackOff(s.ctx, err)
+			continue
+		}
+		s.ln = ln
+		break
+	}
+
+	s.logf("serve listening on %v", s.ap)
+	go s.handleServeListenersAccept(h)
+}
+
+// handleServeListenersAccept accepts connections for the Listener.
+// Calls incoming handler in a new goroutine for each accepted connection.
+func (s *serveListener) handleServeListenersAccept(h func(net.Conn)) {
+	defer s.Close()
+
+	for {
+		conn, err := s.ln.Accept()
+		if err != nil {
+			if s.ctx.Err() == nil {
+				// log only if we're not shutting down
+				s.logf("error accepting serve connection: %v", err)
+			}
+			return
+		}
+		if h != nil {
+			go h(conn)
+		}
+	}
+}
+
+// updateServeTCPPortNetMapAddrListenersLocked starts a net.Listen for configured
+// Serve ports on all the node's addresses.
+// Existing Listeners are closed if port no longer in incoming ports list.
+//
+// b.mu must be held.
+func (b *LocalBackend) updateServeTCPPortNetMapAddrListenersLocked(ports []uint16) {
+	// close existing listeners where port
+	// is no longer in incoming ports list
+	for ap, sl := range b.serveListeners {
+		if !slices.Contains(ports, ap.Port()) {
+			b.logf("closing listener %v", ap)
+			sl.Close()
+			delete(b.serveListeners, ap)
+		}
+	}
+
+	nm := b.netMap
+	if nm == nil {
+		b.logf("netMap is nil")
+		return
+	}
+	if nm.SelfNode == nil {
+		b.logf("netMap SelfNode is nil")
+		return
+	}
+
+	if b.serveListeners == nil {
+		b.serveListeners = make(map[netip.AddrPort]*serveListener)
+	}
+
+	for _, a := range nm.Addresses {
+		for _, p := range ports {
+			addrPort := netip.AddrPortFrom(a.Addr(), p)
+			if _, ok := b.serveListeners[addrPort]; ok {
+				continue // already listening
+			}
+
+			sl := newServeListener(context.Background(), addrPort, b.logf)
```

Comment:
b.newServeListener(...) so it retains b, then line 152 below can just be "go sl.run()"